File: algorithms/1_easy/L/01380_lucky_numbers_in_a_matrix/solution.py
INT_MAX = sys.maxsize
INT_MIN = -sys.maxsize - 1

class Solution:
    def luckyNumbers(self, matrix: List[List[int]]) -> List[int]:
        rows = len(matrix)
        cols = len(matrix[0])

        rows_min_vals = []
        cols_max_vals = []
        res = []

        for j in range(cols):
            max_val = INT_MIN
            for i in range(rows):
                max_val = max(max_val, matrix[i][j])
            cols_max_vals.append(max_val)

        for i in range(rows):
            min_val = INT_MAX
            for j in range(cols):
                min_val = min(min_val, matrix[i][j])
            rows_min_vals.append(min_val)

        for i in range(rows):
            for j in range(cols):
                if matrix[i][j] == rows_min_vals[i] and matrix[i][j] == cols_max_vals[j]:
                    res.append(matrix[i][j])

        return res

File: algorithms/1_easy/L/01380_lucky_numbers_in_a_matrix/solution.cpp
class Solution {
public:
    vector<int> luckyNumbers(vector<vector<int>>& matrix) {
        const size_t rows{matrix.size()};
        const size_t cols{matrix[0].size()};

        vector<int> rows_min_vals;
        rows_min_vals.reserve(rows);
        vector<int> cols_max_vals;
        cols_max_vals.reserve(cols);
        vector<int> res{};

        for (size_t j{0}; j < cols; ++j) {
            int max{INT_MIN};
            for (int i{0}; i < rows; ++i) {
                max = std::max(max, matrix[i][j]);
            }
            cols_max_vals.push_back(max);
        }

        for (size_t i{0}; i < rows; ++i) {
            int min{INT_MAX};
            for (int j{0}; j < cols; ++j) {
                min = std::min(min, matrix[i][j]);
            }
            rows_min_vals.push_back(min);
        }

        for (size_t i{0}; i < rows; ++i) {
            for (size_t j{0}; j < cols; ++j) {
                if (matrix[i][j] == rows_min_vals[i] && matrix[i][j] == cols_max_vals[j]) {
                    res.push_back(matrix[i][j]);
                }
            }
        }

        return res;
    }
};
